File: basic_dsa_one_hundred_one.java
/**
 * Basic DSA Problems - Binary Search
 * Find target in a sorted array using binary search
 */
public class basic_dsa_one_hundred_one {

    // Problem: Given a sorted integer array nums and target, return the index of target.
    // If target is not present, return -1.

    /**
     * Classic binary search: narrow range by comparing mid with target.
     * Time O(log n), Space O(1)
     */
    public static int binarySearch(int[] nums, int target) {
        if (nums == null) return -1;
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) return mid;
            if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Binary Search");
        System.out.println("=".repeat(60));

        int[] a = { -1, 0, 3, 5, 9, 12 };
        System.out.println("nums = [-1,0,3,5,9,12], target = 9 -> " + binarySearch(a, 9));
        System.out.println("nums = [-1,0,3,5,9,12], target = 2 -> " + binarySearch(a, 2));

        int[] b = { 5 };
        System.out.println("nums = [5], target = 5 -> " + binarySearch(b, 5));
        System.out.println("nums = [5], target = -5 -> " + binarySearch(b, -5));
    }
}

File: basic_dsa_one_hundred_one.cpp
/**
 * Basic DSA - Missing Number (LeetCode 268)
 * Find the only number missing from [0, n] in an array of n integers.
 */
#include <iostream>
#include <vector>

int missingNumber(std::vector<int>& nums) {
    int n = nums.size();
    int sum = n * (n + 1) / 2;
    for (int x : nums) sum -= x;
    return sum;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Missing Number\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {3, 0, 1};
    std::cout << "[3,0,1] -> " << missingNumber(a1) << "\n";

    std::vector<int> a2 = {0, 1};
    std::cout << "[0,1] -> " << missingNumber(a2) << "\n";

    std::vector<int> a3 = {9, 6, 4, 2, 3, 5, 7, 0, 1};
    std::cout << "[9,6,4,2,3,5,7,0,1] -> " << missingNumber(a3) << "\n";
}
